interop-testing/src/main/java/io/grpc/testing/integration/TrafficControlProxy.java

```diff
@@ -0,0 +1,319 @@
+package io.grpc.testing.integration;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import com.google.common.base.Preconditions;
+
+import io.netty.channel.socket.nio.NioSocketChannel;
+
+import java.io.DataInputStream;
+import java.io.DataOutputStream;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.lang.reflect.Field;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+// import java.lang.annotation.Native;
+import java.net.InetSocketAddress;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.net.SocketException;
+import java.net.SocketImpl;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.DelayQueue;
+import java.util.concurrent.Delayed;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.logging.Logger;
+
+public final class TrafficControlProxy implements Runnable {
+
+  private static int DEFAULT_BAND_BPS = 1024 * 1024;
+  private static int DEFAULT_DELAY_NANOS = 200 * 1000 * 1000;
+  private static Logger logger = Logger.getLogger(TrafficControlProxy.class.getName());
+
+  // TODO: make ports arguments (and probably loopback/ip as well)
+  private String loopback = "127.0.0.1";
+  private int proxyPort = 5050;
+  private int serverPort = 5001;
+  private int queueLength;
+  private int chunkSize;
+  private int bandwidth;
+  private long latency;
+  private final BlockingQueue<Message> serverQueue = new DelayQueue<Message>();
+  private final BlockingQueue<Message> clientQueue = new DelayQueue<Message>();
+  private volatile int serverQueueSize;
```

Comment:
I don't think `serverQueueSize` needs to be volatile since you always access it under a lock, or after a thread is exited.